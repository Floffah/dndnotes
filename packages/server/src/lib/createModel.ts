import { Schema, model } from "mongoose";

export function createModel<S extends Schema>(name: string, schema: S) {
    schema.virtual("id").get(function (this: any) {
        return this._id.toHexString();
    });

    schema.set("timestamps", true);

    schema.set("toJSON", {
        virtuals: true,
        transform: (_doc: any, ret: any) => {
            delete ret._id;
            delete ret.__v;
        },
    });

    schema.set("toObject", {
        virtuals: true,
        transform: (_doc: any, ret: any) => {
            delete ret._id;
            delete ret.__v;
        },
    });

    return model(name, schema, undefined, {
        overwriteModels: process.env.NODE_ENV === "development",
    });
}
