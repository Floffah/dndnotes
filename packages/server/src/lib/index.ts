export * from "./createBackendCaller";
export * from "./mongoDB";
