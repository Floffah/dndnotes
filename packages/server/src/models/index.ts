export * from "./CampaignInviteModel";
export * from "./CampaignMemberModel";
export * from "./CampaignModel";
export * from "./CampaignSessionModel";
export * from "./CampaignSessionScheduleModel";
export * from "./CharacterModel";
export * from "./DocumentModel";
export * from "./FriendshipRequestModel";
export * from "./UserModel";
export * from "./UserSessionModel";
