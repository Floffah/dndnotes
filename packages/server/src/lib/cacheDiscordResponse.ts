import { REST, RequestData, RequestMethod, RouteLike } from "@discordjs/rest";

import { CachedDiscordResponseModel } from "@/models/CachedDiscordResponseModel";

export async function cacheDiscordResponse<ResponseType>(
    restClient: REST,
    method: RequestMethod,
    route: RouteLike,
    options?: RequestData,
) {
    const existingResponse = await CachedDiscordResponseModel.findOne({
        path: route,
        method: method,
        requestBody: options?.body ?? null,
    });

    if (existingResponse && existingResponse.expiresAt > new Date()) {
        return existingResponse.response as ResponseType;
    }

    let response: unknown;

    const abortController = new AbortController();
    const timeout = setTimeout(() => {
        abortController.abort("Request timed out");
    }, 10000);

    try {
        response = await restClient.request({
            ...options,
            fullRoute: route,
            method: method,
            signal: abortController.signal,
        });
    } catch (e) {
        clearTimeout(timeout);

        if (existingResponse) {
            return existingResponse.response as ResponseType;
        }

        throw e;
    }

    clearTimeout(timeout);

    await CachedDiscordResponseModel.updateOne(
        {
            path: route,
            method: method,
            requestBody: options?.body ?? null,
        },
        {
            path: route,
            method: method,
            requestBody: options?.body ?? null,
            expiresAt: new Date(Date.now() + 1000 * 60 * 60 * 24),
            response: response,
        },
        {
            upsert: true,
        },
    );

    return response as ResponseType;
}
