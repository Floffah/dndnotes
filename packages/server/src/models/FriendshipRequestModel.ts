import { Schema } from "mongoose";

import { FriendshipRequest, FriendshipRequestState } from "@dndnotes/models";

import { createModel } from "@/lib/createModel";

export const FriendshipRequestSchema = new Schema<FriendshipRequest>({
    sender: {
        type: Schema.Types.ObjectId,
        ref: "User",
        required: true,
        index: true,
    },
    recipient: {
        type: Schema.Types.ObjectId,
        ref: "User",
        required: true,
        index: true,
    },
    state: {
        type: String,
        enum: FriendshipRequestState,
        required: true,
        default: FriendshipRequestState.PENDING,
    },
});

export const FriendshipRequestModel = createModel(
    "FriendshipRequest",
    FriendshipRequestSchema,
);
