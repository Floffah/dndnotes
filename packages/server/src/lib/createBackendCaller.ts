import { createRouterCaller } from "@dndnotes/backend-framework/server";

import { appRouter } from "@/router";
import { createContext } from "@/router/context";

export async function createBackendCaller(request: Partial<Request>) {
    const ctx = await createContext({
        req: request as Request,
        resHeaders: new Headers(),
    });

    return createRouterCaller(appRouter, ctx);
}

export type ServerCaller = Awaited<ReturnType<typeof createBackendCaller>>;
