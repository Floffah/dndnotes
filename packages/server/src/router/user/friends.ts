import { ObjectId } from "mongodb";
import { FilterQuery } from "mongoose";
import { z } from "zod";

import { ServerError } from "@dndnotes/backend-framework";
import {
    FriendshipRequestAPIModel,
    FriendshipRequestError,
    FriendshipRequestState,
    UserError,
} from "@dndnotes/models";

import { ensureAuthenticated } from "@/lib/ensureAuthenticated";
import {
    FriendshipRequestModel,
    FriendshipRequestSchema,
} from "@/models/FriendshipRequestModel";
import { UserModel } from "@/models/UserModel";
import { procedure, router } from "@/router/context";

export const userFriendsRouter = router({
    getAccepted: procedure(
        z.object({
            user: z.string().optional(),
        }),
    ).query(async (opts) => {
        await ensureAuthenticated(opts.ctx);

        if (opts.input.user && !ObjectId.isValid(opts.input.user)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        const userId = opts.input.user ?? opts.ctx.session!.user.id;

        const requests = await FriendshipRequestModel.find({
            $or: [
                { sender: new ObjectId(userId) },
                { recipient: new ObjectId(userId) },
            ],
            state: FriendshipRequestState.ACCEPTED,
        })
            .populate("sender")
            .populate("recipient")
            .exec();

        return requests.map(
            (request) =>
                new FriendshipRequestAPIModel(request, {
                    user: opts.ctx.session!.user,
                }),
        );
    }),

    getPending: procedure(
        z.object({
            to: z.string().optional(),
            from: z.string().optional(),
        }),
    ).query(async (opts) => {
        await ensureAuthenticated(opts.ctx);

        if (!opts.input.to && !opts.input.from) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.BAD_INPUT,
            });
        }

        if (opts.input.to && opts.input.to !== opts.ctx.session!.user.id) {
            throw new ServerError({
                code: "FORBIDDEN",
                message: FriendshipRequestError.CANNOT_VIEW_OTHERS_PENDING,
            });
        }

        if (opts.input.to && !ObjectId.isValid(opts.input.to)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        if (opts.input.from && !ObjectId.isValid(opts.input.from)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        if (opts.input.from && opts.input.from !== opts.ctx.session!.user.id) {
            throw new ServerError({
                code: "FORBIDDEN",
                message: FriendshipRequestError.CANNOT_VIEW_OTHERS_PENDING,
            });
        }

        const query: FilterQuery<typeof FriendshipRequestSchema> = {
            state: FriendshipRequestState.PENDING,
        };

        if (opts.input.to) {
            query.recipient = new ObjectId(opts.input.to);
        } else if (opts.input.from) {
            query.sender = new ObjectId(opts.input.from);
        }

        const requests = await FriendshipRequestModel.find(query)
            .populate("sender")
            .populate("recipient")
            .exec();

        return requests.map(
            (request) =>
                new FriendshipRequestAPIModel(request, {
                    user: opts.ctx.session!.user,
                }),
        );
    }),

    sendRequest: procedure(
        z.object({
            to: z.object({
                id: z.string().optional(),
                username: z.string().optional(),
            }),
        }),
    ).mutation(async (opts) => {
        await ensureAuthenticated(opts.ctx);

        if (opts.input.to.id && !ObjectId.isValid(opts.input.to.id)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        if (
            opts.input.to.id === opts.ctx.session!.user.id ||
            opts.input.to.username === opts.ctx.session!.user.name
        ) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.CANNOT_SEND_TO_SELF,
            });
        }

        let recipientId = opts.input.to.id;

        if (!recipientId) {
            const toUser = await UserModel.findOne({
                name: opts.input.to.username,
            });

            if (!toUser) {
                throw new ServerError({
                    code: "NOT_FOUND",
                    message: UserError.NOT_FOUND,
                });
            }

            recipientId = toUser.id;
        }

        const existingRequest = await FriendshipRequestModel.findOne({
            sender: new ObjectId(opts.ctx.session!.user.id),
            recipient: recipientId,
        });

        if (existingRequest) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.REQUEST_ALREADY_SENT,
            });
        }

        const request = new FriendshipRequestModel({
            sender: new ObjectId(opts.ctx.session!.user.id),
            recipient: recipientId,
        });

        await request.populate("sender");
        await request.populate("recipient");

        await request.save();

        return new FriendshipRequestAPIModel(request, {
            user: opts.ctx.session!.user,
        });
    }),

    updateRequest: procedure(
        z.object({
            id: z.string(),
            state: z.nativeEnum(FriendshipRequestState),
        }),
    ).mutation(async (opts) => {
        await ensureAuthenticated(opts.ctx);

        if (!ObjectId.isValid(opts.input.id)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        const request = await FriendshipRequestModel.findById(
            new ObjectId(opts.input.id),
        );

        if (!request) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: FriendshipRequestError.NOT_FOUND,
            });
        }

        if (opts.input.state === FriendshipRequestState.PENDING) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.CANNOT_SET_TO_PENDING,
            });
        }

        if (
            opts.input.state === FriendshipRequestState.ACCEPTED &&
            request.recipient.toString() !== opts.ctx.session!.user.id
        ) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.CANNOT_ACCEPT_OTHERS,
            });
        }

        if (
            opts.input.state === FriendshipRequestState.DENIED &&
            request.recipient.toString() !== opts.ctx.session!.user.id
        ) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.CANNOT_DENY_OTHERS,
            });
        }

        if (
            opts.input.state === FriendshipRequestState.DELETED &&
            request.sender.toString() !== opts.ctx.session!.user.id
        ) {
            throw new ServerError({
                code: "BAD_REQUEST",
                message: FriendshipRequestError.CANNOT_DELETE_OTHERS,
            });
        }

        request.state = opts.input.state;

        if (
            opts.input.state === FriendshipRequestState.DENIED ||
            opts.input.state === FriendshipRequestState.DELETED
        ) {
            await request.deleteOne();
        } else {
            await request.save();
        }

        return new FriendshipRequestAPIModel(request, {
            user: opts.ctx.session!.user,
        });
    }),
});
