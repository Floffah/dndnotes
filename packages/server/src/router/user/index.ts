import { ServerError } from "@dndnotes/backend-framework";
import { UserAPIModel, UserError } from "@dndnotes/models";
import { ObjectId } from "mongodb";
import { z } from "zod";

import { UserModel } from "@/models/UserModel";
import { procedure, router } from "@/router/context";
import { userFriendsRouter } from "@/router/user/friends";

export const userRouter = router({
    friends: userFriendsRouter,

    me: procedure(z.void()).query((opts) => {
        if (!opts.ctx.session) return null;

        return new UserAPIModel(opts.ctx.session.user, {
            user: opts.ctx.session?.user,
        });
    }),
    get: procedure(
        z.object({
            id: z.string(),
        }),
    ).query(async (opts) => {
        const user = await UserModel.findById(new Object(opts.input.id));

        if (opts.input.id && !ObjectId.isValid(opts.input.id)) {
            throw new ServerError({
                code: "NOT_FOUND",
                message: UserError.NOT_FOUND,
            });
        }

        if (!user) return null;

        return new UserAPIModel(user, { user: opts.ctx.session?.user });
    }),
});
