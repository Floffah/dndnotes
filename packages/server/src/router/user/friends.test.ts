import { SESSION_TOKEN } from "@dndnotes/lib";
import {
    FriendshipRequest,
    FriendshipRequestState,
    User,
    UserSession,
} from "@dndnotes/models";

import { ServerCaller, createBackendCaller } from "@/lib/createBackendCaller";
import { FriendshipRequestModel } from "@/models/FriendshipRequestModel";
import { resetDatabase } from "@/tests/utils/mongo";
import { createUser } from "@/tests/utils/user";

let user1: User;
let session1: UserSession;
let user2: User;
let session2: UserSession;

beforeAll(async () => {
    await resetDatabase();

    const { user: _user1, session: _session1 } = await createUser(true);
    user1 = _user1;
    session1 = _session1;

    const { user: _user2, session: _session2 } = await createUser(true);
    user2 = _user2;
    session2 = _session2;
});

describe("No friends", () => {
    let api: ServerCaller;

    beforeAll(async () => {
        api = await createBackendCaller({
            headers: new Headers({
                cookie: `${SESSION_TOKEN}=${session1.token}`,
            }),
        });
    });

    test("User has no pending requests", async () => {
        const friends = await api.user.friends.getPending({
            to: user1.id,
        });

        expect(friends).toEqual([]);
    });

    test("User has no friends", async () => {
        const friends = await api.user.friends.getAccepted({
            user: user1.id,
        });

        expect(friends).toEqual([]);
    });
});

describe("Other user sends friend request", () => {
    let apiUser1: ServerCaller;
    let apiUser2: ServerCaller;
    let request: FriendshipRequest;

    beforeAll(async () => {
        apiUser1 = await createBackendCaller({
            headers: new Headers({
                cookie: `${SESSION_TOKEN}=${session1.token}`,
            }),
        });
        apiUser2 = await createBackendCaller({
            headers: new Headers({
                cookie: `${SESSION_TOKEN}=${session2.token}`,
            }),
        });
    });

    beforeEach(async () => {
        if (request) {
            await FriendshipRequestModel.findByIdAndDelete(request.id);
        }

        request = await apiUser2.user.friends.sendRequest({
            to: {
                id: user1.id,
            },
        });
    });

    test("Accept friend request", async () => {
        await apiUser1.user.friends.updateRequest({
            id: request.id,
            state: FriendshipRequestState.ACCEPTED,
        });

        const friends = await apiUser1.user.friends.getAccepted({
            user: user1.id,
        });

        expect(friends).toHaveLength(1);
        expect(friends[0].sender.id).toEqual(user2.id);
    });

    test("Accept friend request sent to another user", () => {
        return expect(
            apiUser2.user.friends.updateRequest({
                id: request.id,
                state: FriendshipRequestState.ACCEPTED,
            }),
        ).rejects.toMatchSnapshot();
    });

    test("Reject friend request", async () => {
        await apiUser1.user.friends.updateRequest({
            id: request.id,
            state: FriendshipRequestState.DENIED,
        });

        const friends = await apiUser1.user.friends.getAccepted({
            user: user1.id,
        });

        expect(friends).toEqual([]);

        const friendshipRequest = await FriendshipRequestModel.findById(
            request.id,
        );

        expect(friendshipRequest).toBeNull();
    });

    test("Reject friend request sent to another user", () => {
        return expect(
            apiUser2.user.friends.updateRequest({
                id: request.id,
                state: FriendshipRequestState.DENIED,
            }),
        ).rejects.toMatchSnapshot();
    });

    test("Delete friend request", async () => {
        await apiUser2.user.friends.updateRequest({
            id: request.id,
            state: FriendshipRequestState.DELETED,
        });

        const friends = await apiUser1.user.friends.getAccepted({
            user: user1.id,
        });

        expect(friends).toEqual([]);

        const friendshipRequest = await FriendshipRequestModel.findById(
            request.id,
        );

        expect(friendshipRequest).toBeNull();
    });

    test("Delete friend request current user did not send", () => {
        return expect(
            apiUser1.user.friends.updateRequest({
                id: request.id,
                state: FriendshipRequestState.DELETED,
            }),
        ).rejects.toMatchSnapshot();
    });
});
