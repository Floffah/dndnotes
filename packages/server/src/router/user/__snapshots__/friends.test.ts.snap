// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Other user sends friend request Accept friend request sent to another user 1`] = `[ServerError: Cannot accept a request that was not sent to you]`;

exports[`Other user sends friend request Delete friend request current user did not send 1`] = `[ServerError: Cannot delete a request that you did not send]`;

exports[`Other user sends friend request Reject friend request sent to another user 1`] = `[ServerError: Cannot deny a request that was not sent to you]`;
